pkg/infrastructure/azure/azure.go

```diff
@@ -105,9 +112,117 @@ func (p *Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionI
 	if err != nil {
 		return fmt.Errorf("error creating resource group %s: %w", resourceGroupName, err)
 	}
+	resourceGroup, err := resourceGroupsClient.Get(ctx, resourceGroupName, nil)
+	if err != nil {
+		return fmt.Errorf("error getting resource group %s: %w", resourceGroupName, err)
+	}
+
 	logrus.Debugf("ResourceGroup.ID=%s", *resourceGroup.ID)
 	p.ResourceGroupName = resourceGroupName
 
+	// Create user assigned identity
+	userAssignedIdentityName := fmt.Sprintf("%s-identity", in.InfraID)
+	armmsiClientFactory, err := armmsi.NewClientFactory(
+		subscriptionID,
+		tokenCredential,
+		&arm.ClientOptions{
+			ClientOptions: policy.ClientOptions{
+				Cloud: cloudConfiguration,
+			},
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create armmsi client: %w", err)
+	}
+	_, err = armmsiClientFactory.NewUserAssignedIdentitiesClient().CreateOrUpdate(
+		ctx,
+		resourceGroupName,
+		userAssignedIdentityName,
+		armmsi.Identity{
+			Location: ptr.To(platform.Region),
+			Tags:     tags,
+		},
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create user assigned identity %s: %w", userAssignedIdentityName, err)
+	}
+	userAssignedIdentity, err := armmsiClientFactory.NewUserAssignedIdentitiesClient().Get(
```

Comment:
Nit: Noticed that the creation of userAssignedIdentity moved from InfraReady() to PreProvision(). Would it benefit us to move this logic to a method of its own so it can be called from any stage of infrastructure creation?